frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -2527,12 +2527,22 @@ const IR::Node* TypeInference::postorder(IR::Cast* expression) {
                 return expression;
             }
         } else if (auto le = expression->expr->to<IR::ListExpression>()) {
-            if (st->fields.size() == 0 && le->size() == 0) {
-                // Empty structs
+            if (st->fields.size() == le->size()) {
```

Comment:
@mbudiu-vmw This change essentially converts most of the ListExpressions to StructExpressions. This means that midend/backend passes that were expecting a ListExpression might need to also be updated to handle StructExpression. Is there a possibility for the work with ListExpression and StructExpression to somehow be unified? Currently those two classes are not derived from each other, which means they (to some extend) need to be handled separately. The idea is that both of those classes represent pretty similar (if not the same) thing - StructExpression just adds some more structure to the data (NamedExpressions and structType). So what are the possibilities here?

Reply:
Yes, all the midend passes that are part of the open-source repository have been converted to handle StructExpressions.
I don't really see how to unify the two classes. They are relatively simple.
In general, the StructExpression has strictly more information than the ListExpression.
Note that there are some passes, like the tuple elimination pass, which require this conversion to be done again. Tuple elimination converts tuples to structs, and thus tuple initializers, which are ListExpressions, need to change to struct initializers, which are StructExpression.